service/service.go

```
@@ -0,0 +1,98 @@
+package service
+
+import (
+	"context"
+	"github.com/TriggerMail/protos/golang/ratelimit"
+	"github.com/golang/protobuf/ptypes"
+	"golang.org/x/time/rate"
+	"strings"
+	"sync"
+	"time"
+)
+
+type Server struct {
+	mtx     sync.Mutex
+	clients map[string]*clientLimit
+}
+
+type clientLimit struct {
+	limiter           *rate.Limiter
+	nextAvailableTime time.Time
+}
+
+func New() *Server {
+	return &Server{mtx: sync.Mutex{}, clients: make(map[string]*clientLimit)}
+}
+
+func createKey(namespace string, esp string) string {
+	return strings.ToLower(namespace) + "." + strings.ToLower(esp)
+}
+
+func (s *Server) newLimiterForKey(key string, limit int) *clientLimit {
+	// A limiter controls how frequently events are allowed to happen. It implements a "token bucket" of size b
+	// initially full and refilled at a rate of r tokens per second. Every 1/r seconds a new token is added to the
+	// bucket and a token is removed from the bucket. If at any time the number of tokens exceeds the bucket size
+	// then extra tokens are discarded. Since we want each email to be scheduled, in our case the rate limit will
+	// be the bucket size where r=b. The limiter makes the assumption that client applications will chunk
+	// emails so that emailCount is less than or equal to the rate limit.
+	burst := limit
+
+	rateLimit := rate.Limit(limit)
+	limiter := rate.NewLimiter(rateLimit, burst)
+	now := time.Now()
+
+	clientLimitInstance := &clientLimit{limiter: limiter, nextAvailableTime: now}
+
+	s.clients[key] = clientLimitInstance
+	return clientLimitInstance
+
+}
+
+func (s *Server) Lock() {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+}
+
+func makeReservation(s *Server, key string, limit int, emailCount int) time.Time {
```

Why not make this a member function if it operates on the Server itself? This function should also probably return an error as well when reservation.OK() is false that we can propagate back to the client to have them reschedule batches in the future.

This function should have more comments to explain to someone reading the code what is happening as well.